Make certificate generation helpers more flexible

joffrey-bion opened the pull request:
⚠️ Note: I'm making this PR against the `2.2.0-eap` branch because of a discussion with @e5l , but given the size of the changes, I can also change it to `main` instead if you prefer.

**Subsystem**

The subsystem is only the `ktor-network-tls-certificates` helper library to generate certificates (usually for tests).

**Motivation**

The initial issue I have with the current state of this library is that the allowed domains are hardcoded (`localhost` and `127.0.0.1`), thus preventing me from allowing hosts like `host.docker.internal` or `0.0.0.0` or potentially others when generating test certificates for a docker registry mock.

Also, plenty of things are hardcoded in these certificate generation helpers, and there is a lot of code duplication. So I figured I could also polish a bit the API and deduplicate things while I'm at it. 

**Solution**

The point of this PR is to allow for more customizability of the certificate/keystore builder to solve the initial problem. Eventually, the builder became enough to replace most of the bodies of the "all-in-one" `generateCertificate` functions. Those functions could be deprecated in favor of the more flexible `buildKeyStore` API (which is now concise enough to rival with those functions), but I wanted to keep this PR focused as it's already broad enough.

Since more properties are exposed, I also made sure to use standard Kotlin or Java types whenever possible, for instance using `X500Principal` instead of a custom class to represent distinguished names. Speaking of `X500Principal`, I also wrote a builder for it but it's not part of this PR to avoid an excessive amount of code. If you believe this is interesting to have in the lib too, I can make a separate PR for this.

Because there are a lot of changes which include refactoring, I first wrote a bunch of tests to cover all the existing behaviour and ensure I didn't break anything.

Every commit stands on its own and is quite focused, so don't hesitate to review commit-by-commit. The tests and `.api` changes are updated after each commit that requires it, so they can be manipulated independently.

joffrey-bion commented on ktor-network/ktor-network-tls/ktor-network-tls-certificates/jvm/src/io/ktor/network/tls/certificates/builders.kt:
```diff
@@ -101,9 +161,10 @@ public class KeyStoreBuilder internal constructor() {
-            store.setCertificateEntry(alias, certificate)
-            store.setKeyEntry(alias, keys.private, password.toCharArray(), arrayOf(certificate))
+            val (certificate, keys, password, issuerCertificate) = info
+            store.setCertificateEntry(alias + "Cert", certificate)
```
A note on this change. `setCertificateEntry` uses the same hashtable internally as `setKeyEntry`, which means calling `setKeyEntry` with the same alias makes `setCertificateEntry` redundant - we could remove this line entirely.

However, one of the `generateCertificate` functions adds the certificate to the keystore using this `Cert` suffix, so to ensure backwards compatibility with existing usages of `generateCertificate`, I added this suffix here too instead of removing the line. If you believe it's not worth keeping it backwards compatible to this extent, we can remove the line and the corresponding test.

joffrey-bion replied:
It turns out the tests are failing because this change is backwards-incompatible for the other code paths: some tests iterate on all aliases in the keystore and they expect to get a certificate chain out of every alias, which only works for aliases on private keys, not for aliases of plain certificates.

I will remove this line entirely here, and only add the certificate under the `<keyAlias>Cert` alias in the case of the `generateCertificate` function that used to do it. It should be possible by accessing the generated certificate from the key store and then re-adding it under the other alias.

joffrey-bion replied:
Done.

joffrey-bion commented on ktor-network/ktor-network-tls/ktor-network-tls-certificates/jvm/src/io/ktor/network/tls/certificates/builders.kt:
```diff
@@ -47,38 +57,88 @@ public class CertificateBuilder internal constructor() {
+     * If this method is not called, this certificate will be self-signed by the subject (with the same generated key
+     * as the one being certified).
+     */
+    public fun signWith(
+        issuerKeyPair: KeyPair,
+        issuerKeyCertificate: Certificate,
+        issuerName: X500Principal,
+    ) {
```
This could also just be named `issuer`, with params like `issuer(name: X500Principal, keyPair: KeyPair, keyCertificate: Certificate)`. I just felt like the name `issuer` feels weird for the other overload where the name of the issuer is taken from the certificate. Maybe we don't need that other overload at all 🤷‍♂️

joffrey-bion commented:
☝️ fixed the line that was too long

e5l reviewed (changes requested):
LGTM in general, please check minor comments and failing tests

joffrey-bion commented on ktor-network/ktor-network-tls/ktor-network-tls-certificates/jvm/src/io/ktor/network/tls/certificates/Certificates.kt:
```diff
@@ -14,8 +15,16 @@ import java.security.cert.*
+import kotlin.time.Duration.Companion.days
+
+private val jbLocalhostPrincipal = X500Principal("CN=localhost, OU=Kotlin, O=JetBrains, C=RU")
+private val jbLocalhostCAPrincipal = X500Principal("CN=localhostCA, OU=Kotlin, O=JetBrains, C=RU")
```
I agree it should be Ktor, but it was `Kotlin` in the existing code, so I kept it this way. This could be a backwards-incompatible change

e5l replied:
Agree, let's leep it

joffrey-bion commented on ktor-network/ktor-network-tls/ktor-network-tls-certificates/jvm/src/io/ktor/network/tls/certificates/Certificates.kt:
```diff
@@ -35,69 +44,70 @@ public fun generateCertificate(
+            this.keySizeInBits = keySizeInBits
+            this.keyType = keyType
+            this.subject = if (keyType == KeyType.CA) jbLocalhostCAPrincipal else jbLocalhostPrincipal
+            this.domains = listOf("127.0.0.1", "localhost")
```
We need it for `keySizeInBits` and for `keyType`, and I wanted to align other properties visually. If you think it's not worth the alignment, I can remove the `this.` for other properties.

e5l replied:
it's ok, but looks strange

joffrey-bion commented:
☝️ the above force-push is for little tweaks:

* added a test in the initial `Add tests for certificate generation` commit to verify the exact aliases generated in the keyStore (to ensure we don't generate less or more aliases)
* fixed the CertificateBuilder generalization commit (the last one) so we are 100% backwards-compatible with the aliases generated: the keystore builder and CA-based `KeyStore.generateCertificate` function only add one alias, and the top-level `generateCertificate` function still generates 2 aliases (one with `Cert` suffix)

e5l reviewed (approved):
lgtm
